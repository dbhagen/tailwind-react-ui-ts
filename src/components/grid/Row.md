```jsx
<Row>
  <Col bg="gray-400" text="center" p="4">
    1/1
  </Col>
  <Col w="1/2" bg="gray-300" text="center" p="4">
    1/2
  </Col>
  <Col w="1/2" bg="gray-400" text="center" p="4">
    1/2
  </Col>
  <Col w="1/3" bg="gray-300" text="center" p="4">
    1/3
  </Col>
  <Col w="1/3" bg="gray-400" text="center" p="4">
    1/3
  </Col>
  <Col w="1/3" bg="gray-300" text="center" p="4">
    1/3
  </Col>
  <Col w="1/4" bg="gray-400" text="center" p="4">
    1/4
  </Col>
  <Col w="1/4" bg="gray-300" text="center" p="4">
    1/4
  </Col>
  <Col w="1/4" bg="gray-400" text="center" p="4">
    1/4
  </Col>
  <Col w="1/4" bg="gray-300" text="center" p="4">
    1/4
  </Col>
</Row>
```

### Gutters

```jsx
<Row gutter>
  <Col>
    <Card bg="gray-400" text="center" p="4" rounded="none">
      1/1
    </Card>
  </Col>
  <Col w="1/2">
    <Card bg="gray-300" text="center" p="4" rounded="none">
      1/2
    </Card>
  </Col>
  <Col w="1/2">
    <Card bg="gray-400" text="center" p="4" rounded="none">
      1/2
    </Card>
  </Col>
  <Col w="1/3">
    <Card bg="gray-300" text="center" p="4" rounded="none">
      1/3
    </Card>
  </Col>
  <Col w="1/3">
    <Card bg="gray-400" text="center" p="4" rounded="none">
      1/3
    </Card>
  </Col>
  <Col w="1/3">
    <Card bg="gray-300" text="center" p="4" rounded="none">
      1/3
    </Card>
  </Col>
  <Col w="1/4">
    <Card bg="gray-400" text="center" p="4" rounded="none">
      1/4
    </Card>
  </Col>
  <Col w="1/4">
    <Card bg="gray-300" text="center" p="4" rounded="none">
      1/4
    </Card>
  </Col>
  <Col w="1/4">
    <Card bg="gray-400" text="center" p="4" rounded="none">
      1/4
    </Card>
  </Col>
  <Col w="1/4">
    <Card bg="gray-300" text="center" p="4" rounded="none">
      1/4
    </Card>
  </Col>
</Row>
```

### Responsive Columns

```jsx
<Row gutter>
  <Col w={{ def: 'full', sm: '1/2', md: '1/3', lg: '1/4', xl: '1/6' }}>
    <Card bg="gray-400" p={4} h={12} rounded="none" />
  </Col>
  <Col w={{ def: 'full', sm: '1/2', md: '1/3', lg: '1/4', xl: '1/6' }}>
    <Card bg="gray-300" p={4} h={12} rounded="none" />
  </Col>
  <Col w={{ def: 'full', sm: '1/2', md: '1/3', lg: '1/4', xl: '1/6' }}>
    <Card bg="gray-400" p={4} h={12} rounded="none" />
  </Col>
  <Col w={{ def: 'full', sm: '1/2', md: '1/3', lg: '1/4', xl: '1/6' }}>
    <Card bg="gray-300" p={4} h={12} rounded="none" />
  </Col>
  <Col w={{ def: 'full', sm: '1/2', md: '1/3', lg: '1/4', xl: '1/6' }}>
    <Card bg="gray-400" p={4} h={12} rounded="none" />
  </Col>
  <Col w={{ def: 'full', sm: '1/2', md: '1/3', lg: '1/4', xl: '1/6' }}>
    <Card bg="gray-300" p={4} h={12} rounded="none" />
  </Col>
</Row>
```

### Auto Sized Columns

```jsx
<Row nowrap>
  <Col w="auto">
    <Card bg="gray-400" p={4} h={12} rounded="none" />
  </Col>
  <Col w="auto">
    <Card bg="gray-300" p={4} h={12} rounded="none" />
  </Col>
  <Col w="auto">
    <Card bg="gray-400" p={4} h={12} rounded="none" />
  </Col>
</Row>
```
