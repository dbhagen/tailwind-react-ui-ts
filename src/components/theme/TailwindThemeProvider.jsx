import React from 'react'
import PropTypes from 'prop-types'
import merge from 'lodash.merge'

import defaultTheme from './defaultTheme'
import TailwindTheme from './TailwindTheme'

const TailwindThemeProvider = ({ theme, children }) => {
  const mergedTheme = merge({}, defaultTheme, theme)

  return (
    <TailwindTheme.Provider value={mergedTheme}>
      {children}
    </TailwindTheme.Provider>
  )
}

TailwindThemeProvider.propTypes = {
  theme: PropTypes.shape({}),
  children: PropTypes.node,
}

TailwindThemeProvider.defaultProps = {
  theme: {},
  children: undefined,
}

export default TailwindThemeProvider
