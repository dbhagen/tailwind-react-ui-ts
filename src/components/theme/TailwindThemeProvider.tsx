import React from 'react'
import merge from 'lodash.merge'

import defaultTheme from './defaultTheme'
import TailwindTheme from './TailwindTheme'

export type TailwindThemeProviderProps = {
  theme: {}
  children: React.ReactNode
}

const TailwindThemeProvider: React.FunctionComponent<TailwindThemeProviderProps> = ({ theme, children }) => {
  const mergedTheme = merge({}, defaultTheme, theme)

  return (
    <TailwindTheme.Provider value={mergedTheme}>
      {children}
    </TailwindTheme.Provider>
  )
}

export default TailwindThemeProvider
