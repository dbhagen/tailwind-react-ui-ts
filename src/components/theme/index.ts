export { default as defaultTheme } from './defaultTheme'
export { default as TailwindTheme } from './TailwindTheme'
export { default as TailwindThemeProvider } from './TailwindThemeProvider'
export { default as withTheme } from './withTheme'
export { default as useThemeValue } from './useThemeValue'
