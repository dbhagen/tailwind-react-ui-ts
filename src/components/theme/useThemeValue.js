import { tailwindPropToClassName } from '../tailwind'

export default (prefix, value, userClassNames = '') => {
  if (userClassNames.includes(prefix)) {
    return false
  }

  return tailwindPropToClassName(prefix, value)
}
