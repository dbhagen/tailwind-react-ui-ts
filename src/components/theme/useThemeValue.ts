import { tailwindPropToClassName } from '../tailwind'

export default (prefix: string, values: Object, userClassNames = '') => {
  if (userClassNames.includes(prefix)) {
    return false
  }

  return tailwindPropToClassName(undefined, values, prefix)
}
