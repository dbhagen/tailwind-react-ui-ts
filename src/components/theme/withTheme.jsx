import React from 'react'

import TailwindTheme from './TailwindTheme'

const withTheme = Component => {
  const WithTheme = props => (
    <TailwindTheme.Consumer>
      {theme => <Component {...props} theme={theme} />}
    </TailwindTheme.Consumer>
  )

  WithTheme.displayName = `WithTheme(${Component.displayName})`

  return WithTheme
}

export default withTheme
