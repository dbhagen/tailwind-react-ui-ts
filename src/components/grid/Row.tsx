import React from 'react'
import PropTypes from 'prop-types'

import { withTheme } from '../theme'
import { Flex } from '../primitives'

const Row = ({ is, children, nowrap, gutter, theme, ...rest }) => {
  const gutterSpacing =
    gutter && (gutter === true ? theme.spacing.md : theme.spacing[gutter])

  return (
    <Flex
      is={is}
      wrap={!nowrap}
      nm={
        gutter
          ? { l: gutterSpacing, b: !nowrap ? gutterSpacing : undefined }
          : undefined
      }
      listReset
      {...rest}
    >
      {gutter
        ? React.Children.map(children, child =>
            React.cloneElement(child, {
              p: { l: gutterSpacing },
              m: { b: gutterSpacing },
            }),
          )
        : children}
    </Flex>
  )
}

Row.propTypes = {
  theme: PropTypes.shape({}).isRequired,
  is: PropTypes.oneOfType([PropTypes.string, PropTypes.func, PropTypes.object]),
  children: PropTypes.node,
  nowrap: PropTypes.bool,
  gutter: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]),
}

Row.defaultProps = {
  is: 'ul',
  children: undefined,
  nowrap: false,
  gutter: false,
}

export { Row as component }
export default withTheme(Row)
