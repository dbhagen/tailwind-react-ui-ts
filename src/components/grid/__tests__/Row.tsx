import React from 'react'
import { shallow } from 'enzyme'

import { component as Row } from '../Row'
import { defaultTheme } from '../../theme'

const setup = (testProps = {}) => {
  const props = Object.assign({ theme: defaultTheme }, testProps)

  const wrapper = shallow(<Row {...props} />)

  return {
    props,
    wrapper,
  }
}

describe('Row', () => {
  it('renders matching snapshot', () => {
    const { wrapper } = setup({ gutter: true })

    expect(wrapper).toMatchSnapshot()
  })
})
