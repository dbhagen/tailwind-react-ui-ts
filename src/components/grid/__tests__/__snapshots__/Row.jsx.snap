// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Row renders matching snapshot 1`] = `
<Flex
  col={false}
  inline={false}
  inlineFlex={false}
  is="ul"
  listReset={true}
  nm={
    Object {
      "b": 4,
      "l": 4,
    }
  }
  reverse={false}
  wrap={true}
  wrapReverse={false}
/>
`;
